[InteractionRegions] Highlight container instead of shape when it can improve visibility

etiennesegonzac opened the pull request:
#### 16668679e31ce64c7c8cee5d7d64e2c4b7587400
<pre>
[InteractionRegions] Highlight container instead of shape when it can improve visibility
<a href="https://bugs.webkit.org/show_bug.cgi?id=275208">https://bugs.webkit.org/show_bug.cgi?id=275208</a>
&lt;<a href="https://rdar.apple.com/128509643">rdar://128509643</a>&gt;

Reviewed by Simon Fraser.

Detect cases where the highlight doesn&apos;t work well (small shapes with
very high or very low luminance) and fallback to highlighting their
containers instead.

In order to do so we go over Guards and convert them to Interactions if
the Interaction they&apos;re guarding doesn&apos;t exist.

* Source/WebCore/page/InteractionRegion.cpp:
(WebCore::usesFillColorWithExtremeLuminance):
New function to detect fill colors that won&apos;t work well.
(WebCore::interactionRegionForRenderedRegion):
Don&apos;t generate an Interaction when it won&apos;t be visible enough and a
container is available.

* Source/WebCore/rendering/EventRegion.h:
* Source/WebCore/rendering/EventRegion.cpp:
(WebCore::EventRegionContext::convertGuardContainersToInterationIfNeeded):
New method to convert Guards to Interactions when needed. It needs to
know the minimum corner radius to apply it.
Guards containers have &quot;transparent&quot; style so it&apos;s always safe to add.
(WebCore::EventRegionContext::copyInteractionRegionsToEventRegion):
Call the new method.

* Source/WebCore/rendering/RenderLayerBacking.cpp:
(WebCore::RenderLayerBacking::updateEventRegion):
* Source/WebCore/rendering/RenderLayerCompositor.cpp:
(WebCore::RenderLayerCompositor::updateScrollLayerClipping):
Pass the minimum corner radius setting down.

* LayoutTests/interaction-region/svg-luminance-expected.txt: Added.
* LayoutTests/interaction-region/svg-luminance.html: Added.
Add new tests to cover this change.

Canonical link: <a href="https://commits.webkit.org/279859@main">https://commits.webkit.org/279859@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/4cb69dad432a481ce5532c3f61289b2e64c23cb7

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/54763 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/34191 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/7339 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/58042 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/5495 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/57064 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/41738 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/5508 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/44354 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/3711 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/56858 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/32323 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/47441 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/25478 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/29113 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/4779 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/3636 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/50891 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/4996 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/59632 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/30009 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/5141 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/51777 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/31151 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/47524 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/51179 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/12027 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/32158 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/30938 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

smfr commented on Source/WebCore/rendering/EventRegion.h:
```diff
@@ -62,9 +62,10 @@ class EventRegionContext final : public RegionContext {
     void removeSuperfluousInteractionRegions();
     void shrinkWrapInteractionRegions();
-    void copyInteractionRegionsToEventRegion();
+    void copyInteractionRegionsToEventRegion(float);
```
The argument needs a name.

smfr commented on Source/WebCore/page/InteractionRegion.cpp:
```diff
@@ -208,6 +208,21 @@ static bool hasTransparentContainerStyle(const RenderStyle& style)
 
+static bool usesFillColorWithExtremeLuminance(const RenderStyle& style)
+{
+    auto fillColor = style.fillPaintColor();
```
This should use `style().colorResolvingCurrentColor()` instead of the manual current color resolution below.

etiennesegonzac replied:
nice, adding `currentcolor` test cases too

smfr commented on LayoutTests/interaction-region/svg-lumincance-expected.txt:
```diff
@@ -0,0 +1,41 @@
+      (event region
+        (rect (0,0) width=800 height=600)
+
+      (interaction regions [
```
We should invent a ref test mode that actually paints these regions in the result. Diffing/reviewing these paths is next to impossible.

etiennesegonzac replied:
excellent idea, making sure it's captured!

nt1m commented on LayoutTests/interaction-region/svg-lumincance.html:
Nit: typo in filename (lumincance)

etiennesegonzac committed:
[InteractionRegions] Highlight container instead of shape when it can improve visibility
https://bugs.webkit.org/show_bug.cgi?id=275208
<rdar://128509643>

Reviewed by Simon Fraser.

Detect cases where the highlight doesn't work well (small shapes with
very high or very low luminance) and fallback to highlighting their
containers instead.

In order to do so we go over Guards and convert them to Interactions if
the Interaction they're guarding doesn't exist.

* Source/WebCore/page/InteractionRegion.cpp:
(WebCore::usesFillColorWithExtremeLuminance):
New function to detect fill colors that won't work well.
(WebCore::interactionRegionForRenderedRegion):
Don't generate an Interaction when it won't be visible enough and a
container is available.

* Source/WebCore/rendering/EventRegion.h:
* Source/WebCore/rendering/EventRegion.cpp:
(WebCore::EventRegionContext::convertGuardContainersToInterationIfNeeded):
New method to convert Guards to Interactions when needed. It needs to
know the minimum corner radius to apply it.
Guards containers have "transparent" style so it's always safe to add.
(WebCore::EventRegionContext::copyInteractionRegionsToEventRegion):
Call the new method.

* Source/WebCore/rendering/RenderLayerBacking.cpp:
(WebCore::RenderLayerBacking::updateEventRegion):
* Source/WebCore/rendering/RenderLayerCompositor.cpp:
(WebCore::RenderLayerCompositor::updateScrollLayerClipping):
Pass the minimum corner radius setting down.

* LayoutTests/interaction-region/svg-luminance-expected.txt: Added.
* LayoutTests/interaction-region/svg-luminance.html: Added.
Add new tests to cover this change.

Canonical link: https://commits.webkit.org/279859@main
